add toggle in `mr-app` for `preserve-drawing-buffer`

hanbollar opened the pull request:
## Linking

Fixes https://github.com/Volumetrics-io/mrjs/issues/506

## Problem

Sometimes it's nice to be able to right click the canvas, select "Save image as" and get a PNG of the MRjs scene, with preserved transparency.

## Solution

add in an ability to toggle the `preserve-drawing-buffer` flag for the threejs renderer

### Breaking Change

adds the `<mr-app>` flag `preserve-drawing-buffer`

## Notes

            // There's issues in the timing to enable taking screenshots of threejs scenes unless you have direct access to the code.
            // Using the preserveDrawingBuffer to ignore timing issues is the best approach instead. Though this has a performance hit,
            // we're allowing it to be enabled by users when necessary.
            //
            // References:
            // https://stackoverflow.com/questions/15558418/how-do-you-save-an-image-from-a-three-js-canvas
            // https://stackoverflow.com/questions/30628064/how-to-toggle-preservedrawingbuffer-in-three-js

------------

## Required to Merge

- [x] **PASS** - all necessary actions must pass (excluding the auto-skipped ones)
~~- [ ] **TEST IN HEADSET** - [main dev-testing-example](https://github.com/Volumetrics-io/mrjs/tree/main/samples/index.html) and any of the other [examples](https://github.com/Volumetrics-io/mrjs/tree/main/samples/examples) still work as expected~~
~~- [ ] **VIDEO** - if this pr changes something visually - post a video here of it in headset-MR and/or on desktop (depending on what it affects) for the reviewer to reference.~~
- [x] **TITLE** - make sure the pr's title is updated appropriately as it will be used to name the commit on merge
- [x] **BREAKING CHANGE**
  - make a pr in the [documentation repo](https://github.com/Volumetrics-io/documentation) that updates the manual docs to match the breaking change
  - note: the docs underneath the `Javascript API` section come automated from the jsdocs inline with the code itself
  - link the pr of the documentation repo here: *https://github.com/Volumetrics-io/documentation/pull/26*
  - that documentation repo pr must be approved by `@lobau`

takahirox commented on src/core/MRApp.js:
```diff
@@ -202,6 +202,16 @@ export class MRApp extends MRElement {
+            // References:
+            // https://stackoverflow.com/questions/15558418/how-do-you-save-an-image-from-a-three-js-canvas
+            // https://stackoverflow.com/questions/30628064/how-to-toggle-preservedrawingbuffer-in-three-js
+            this.renderer.preserveDrawingBuffer = true;
```
`preserveDrawingBuffer` flag must be passed via the `WebGLRenderer` constructor.

https://threejs.org/docs/#api/en/renderers/WebGLRenderer

hanbollar replied:
good catch with that

---

very odd, i feel like im not seeing any difference using it in constructor, outside, or even not having it at all

i wonder if it's a browser thing - i'll look into this and follow back up for why the discrepancy as we might not even need the flag then if it works?

weird...

hanbollar replied:
discussed - following up per [conversation](https://github.com/Volumetrics-io/mrjs/issues/506#issuecomment-2007921050) for why issue wasnt showing up on my end - and [added note about chromium base](https://github.com/Volumetrics-io/documentation/pull/26/commits/9f623225d10f8ad7e0d99e0e815032ab8bca2dde) to the documentation pr

hanbollar replied:
going to continue to implement a working fix ~

hanbollar commented:
recap of fix since the last review
- moved s.t. `preserve-drawing-buffer` is properly in the constructor --> had to move renderer creation to do so, but works fine with that setup
- added a debug.html bc it got annoying adding and removing the flags over and over --> this is a small thing for now, in future we'll probably make a more robust html for it
- added the html to testing
- updated the documentation pr to reflect the browser note as well (regarding need or no-need for the flag itself)

@takahirox @michaelthatsit lmk if there's anything else that might be blocking for this pr being merged ~

michaelthatsit commented on src/core/MRApp.js:
```diff
@@ -193,6 +195,19 @@ export class MRApp extends MRElement {
+            // References:
+            // https://stackoverflow.com/questions/15558418/how-do-you-save-an-image-from-a-three-js-canvas
+            // https://stackoverflow.com/questions/30628064/how-to-toggle-preservedrawingbuffer-in-three-js
+            preserveDrawingBuffer: this.getAttribute('preserve-drawing-buffer') ?? false,
```
Maybe we shorten it to `preserve-buffer`? Could just be my aversion to long attribute names.

hanbollar replied:
I agree with the reasoning; however, I'd rather not given this is us just exposing a threejs flag underneath so it's better practice for maintainability to keep them the same - if it was one we made ourselves though 100% shorter is better

michaelthatsit replied:
Fair point. maybe in the future as we expose more renderer options at the attribute level, we can fall back to an all encompassing `data-renderer` flag. something like.

`<mr-app data-renderer="antialias: true; alpha: true; preserveDrawingBuffer: true; ...">

rather than having one for each possible value.

hanbollar replied:
i like that idea - i'll open an issue for tracking and it'll probably stay on the backlog for a bit, but there's definitely other items in there we'd want to expose as well ~

hanbollar replied:
issue ref: #519
